pkg/obs/nodeserver.go

```diff
@@ -17,38 +17,147 @@ limitations under the License.
 }
 
+const (
+	credentialFile = "/host/etc/passwd-obsfs"
+	credential     = "/etc/passwd-obsfs"
+	defaultOpts    = "-o big_writes -o max_write=131072 -o use_ino"
+	perm           = 0600
+)
+
 func (ns *nodeServer) NodeStageVolume(_ context.Context, req *csi.NodeStageVolumeRequest) (
 	*csi.NodeStageVolumeResponse, error) {
+	log.Infof("NodeStageVolume: called with args %v", protosanitizer.StripSecrets(*req))
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
 func (ns *nodeServer) NodeUnstageVolume(_ context.Context, req *csi.NodeUnstageVolumeRequest) (
 	*csi.NodeUnstageVolumeResponse, error) {
+	log.Infof("NodeStageVolume: called with args %v", protosanitizer.StripSecrets(*req))
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
 func (ns *nodeServer) NodePublishVolume(_ context.Context, req *csi.NodePublishVolumeRequest) (
 	*csi.NodePublishVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	log.Infof("NodePublishVolume: called with args %v", protosanitizer.StripSecrets(*req))
+	capability := req.GetVolumeCapability()
+	volumeID := req.GetVolumeId()
+	targetPath := req.GetTargetPath()
+	if err := nodePublishValidation(capability, volumeID, targetPath); err != nil {
+		return nil, err
+	}
+
+	credentials := ns.Driver.cloud
+	volume, err := services.GetParallelFSBucket(credentials, volumeID)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("NodePublishVolume: volume detail: %s", protosanitizer.StripSecrets(volume))
+
+	notMnt, err := ns.Mount.IsLikelyNotMountPointAttach(targetPath)
+	if err != nil {
+		return nil, err
+	}
+	if !notMnt {
+		log.Infof("NodePublishVolume: %s has already mounted", targetPath)
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
+	if err := ns.Mount.MakeDir(targetPath); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to make dir: %s, error: %v", targetPath, err)
+	}
+	if err := createCredentialFile(ns.Driver.cloud.Global.AccessKey, ns.Driver.cloud.Global.SecretKey); err != nil {
+		return nil, err
+	}
+
+	mntCmd := fmt.Sprintf("obsfs %s %s -o url=obs.%s.%s -o passwd_file=%s %s", volume.BucketName, targetPath,
+		ns.Driver.cloud.Global.Region, ns.Driver.cloud.Global.Cloud, credential, defaultOpts)
+	log.Infof("NodePublishVolume: obsfs cmd : %s", mntCmd)
+
+	if err := sendCommand(mntCmd, "http://unix/mount", ns.MountClient); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to mount %s at %s: %v",
+			volume.BucketName, targetPath, err)
+	}
+	log.Infof("NodePublishVolume: mount %s at %s successfully", volume.BucketName, targetPath)
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func createCredentialFile(accessKey, secretKey string) error {
+	if err := os.MkdirAll(path.Dir(credentialFile), os.ModePerm); err != nil {
+		return status.Errorf(codes.Internal, "Failed to make dir: %s, error: %v", credentialFile, err)
+	}
+
+	writer, err := os.OpenFile(filepath.Clean(credentialFile), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, perm)
```

Comment:
Don't flush every time, this will cause errors in threads that are in use.

Reply:
this passfile will be deleted by hosting environment,
this file would create every time.